src/rabbit_direct.erl

```diff
@@ -211,6 +211,7 @@ start_channel(Number, ClientChannelPid, ConnPid, ConnName, Protocol, User,
           rabbit_direct_client_sup,
           [{direct, Number, ClientChannelPid, ConnPid, ConnName, Protocol,
             User, VHost, Capabilities, Collector}]),
+    _ = rabbit_channel:source(ChannelPid, ?MODULE),
```

Comment:
Rather than do this, check out `rabbit_channel_sup:start_link`. It differentiates between `tcp` (i.e. `rabbit_reader`) and `direct` which could be passed as an argument to `rabbit_channel:start_link`.

Reply:
i've added `network_arbitrary_channel_source` test case to illustrate why relying on the supervisor for setting the channel source may be a problem. `rabbit_channel_sup:start_link` may still be executed by any entity. Network/tcp based parent connections may still be `amqp_params` info compliant. e.g. `amqp_gen_connection` based connections, and if passed as part of a channel's start args (even for test purposes), then it should still work as expected. have a look at the test case

Reply:
Those tests clarify things. I still am concerned that at some point the call to `rabbit_channel:source/2` will be missed, and the resulting bug hard to diagnose. I'm not sure how likely that scenario is (@michaelklishin ?)

Could `?MODULE` be added as an argument to `start_link`, or an additional argument like `Opts` that would specify `amqp_params_supported`?

Reply:
There are only so many places where channels are started. We can make the parameter required with a shim function head that passes `?MODULE` for it if it's not explicitly provided. These are technically `rabbit_channel` API changes but that's fine since this is node local state.

Reply:
with such a shim function in the channel, the preprocessor will always expand `?MODULE` to `rabbit_channel`, causing a reflection something that's incorrect in state. Rather the source has to explicitly provided. And announcing source isn't mandatory. If `rabbit_channel:source/2` is missed, `source` defaults/remains as `undefined` in state and [is always ignored](https://github.com/rabbitmq/rabbitmq-server/pull/1886/files#diff-98224c82bfd7f4172da6a5adff2c42b0R1027) for this purpose. Unless we're anticipating another **critical** connection implementation aside `rabbit_reader` to consider?

Reply:
Thanks for clarifying. We don't expect any new connection implementations to appear any time soon :)

@lukebakken should I continue with QA or would you like to see more changes?

Reply:
I'm fine